prometheus/src/main/scala/filodb/prometheus/ast/Functions.scala

```diff
@@ -21,6 +21,199 @@ trait Functions extends Base with Operators with Vectors {
+
+    // error messages
+    val errExpectedRangeVector : String = "Expected type range_vector in call to function "
+    val errExpectedScalar : String = "Expected type scalar in call to function "
+    val errExpectedInstantVector : String = "Expected type instant vector in call to function "
+    val errExpected1Argument : String = "Expected 1 argument(s) in call to function "
+    val errExpected2Argument : String = "Expected 2 argument(s) in call to function "
+    val errExpected3Argument : String = "Expected 3 argument(s) in call to function "
+
+    /**
+     *
+     * Below function validates if the first argument to all the functions in RangeFunctionId which needs only one argument as range-expression
+     * is correct or not. If not, then throw Illegal Argument Exception with the proper error message.
+     */
+    def onlyRangeExprParamValidator(funcName : String) : Unit = {
+      // needs to have 1 parameter
+      if (allParams.size == 1) {
+        // 1st Param should be a range-vector
+        if (!allParams.head.isInstanceOf[RangeExpression])
+          throw new IllegalArgumentException(s"${errExpectedRangeVector} ${funcName}, got ${allParams.head.getClass.getSimpleName}")
+      } else {
+        throw new IllegalArgumentException(s"${errExpected1Argument} ${funcName}, got ${allParams.size}")
+      }
+    }
+
+    /**
+     *
+     * Below function validates if the first argument to all the functions in RangeFunctionId which needs only one argument as instant-expression
+     * is correct or not. If not, then throw Illegal Argument Exception with the proper error message.
+     */
+    def onlyInstantExprParamValidator(funcName : String) : Unit = {
+      // needs to have 1 parameter
+      if (allParams.size == 1) {
+        // 1st Param should be a instant-vector
+        if (!allParams.head.isInstanceOf[InstantExpression])
+          throw new IllegalArgumentException(s"${errExpectedInstantVector} ${funcName}, got ${allParams.head.getClass.getSimpleName}")
+      } else {
+        throw new IllegalArgumentException(s"${errExpected1Argument} ${funcName}, got ${allParams.size}")
+      }
+    }
+
+    /**
+     *
+     * holt-winters params validator
+     */
+    def holtWintersParamValidator(funcName : String) : Unit = {
+      // needs to have 3 mandatory params
+      if (allParams.size == 3) {
+
+        // 1st Param should be a range-vector
+        if (!allParams.head.isInstanceOf[RangeExpression])
+          throw new IllegalArgumentException(s"${errExpectedRangeVector} ${funcName}, got ${allParams.head.getClass.getSimpleName}")
+
+        // 2nd param needs to be a scalar
+        if (allParams(1).isInstanceOf[ScalarExpression]) {
+
+          // if it's a scalar, smoothing value needs to be between greater than 0 and less than 1
+          if(!(allParams(1).asInstanceOf[ScalarExpression].toScalar > 0 && allParams(1).asInstanceOf[ScalarExpression].toScalar < 1))
```

Comment:
Code like this can be made more readable by factoring out `allParams(1).asInstanceOf[ScalarExpression].toScalar` to a val just above this.